pkg/dockerregistry/server/pullthroughblobstore.go

```diff
@@ -0,0 +1,204 @@
+		// TODO: normalize further?
+		localRegistry = local.Registry
+	}
+
+	retriever := r.repo.importContext()
+	cached := r.repo.cachedLayers.ReposForLayer(dgst)
+
+	// look at the first level of tagged repositories first
+	search := identifyCandidateRepositories(is, localRegistry, true)
+	if desc, err := r.findCandidateRepository(ctx, search, cached, dgst, retriever); err == nil {
+		return desc, nil
+	}
+
+	// look at all other repositories tagged by the server
+	secondary := identifyCandidateRepositories(is, localRegistry, false)
+	for k := range search {
+		delete(secondary, k)
+	}
+	if desc, err := r.findCandidateRepository(ctx, secondary, cached, dgst, retriever); err == nil {
+		return desc, nil
+	}
+
+	return distribution.Descriptor{}, distribution.ErrBlobUnknown
+}
+
+// proxyStat attempts to locate the digest in the provided remote repository or returns an error. If the digest is found,
+// r.digestToStore saves the store.
+func (r *pullthroughBlobStore) proxyStat(ctx context.Context, retriever importer.RepositoryRetriever, ref imageapi.DockerImageReference, dgst digest.Digest) (distribution.Descriptor, error) {
+	context.GetLogger(r.repo.ctx).Infof("Trying to stat %q from %q", dgst, ref.Exact())
+	repo, err := retriever.Repository(ctx, ref.RegistryURL(), ref.RepositoryName(), false)
+	if err != nil {
+		context.GetLogger(r.repo.ctx).Errorf("Error getting remote repository for image %q: %v", ref.Exact(), err)
+		return distribution.Descriptor{}, err
+	}
+	pullthroughBlobStore := repo.Blobs(ctx)
+	desc, err := pullthroughBlobStore.Stat(r.repo.ctx, dgst)
+	if err != nil {
+		if err != distribution.ErrBlobUnknown {
+			context.GetLogger(r.repo.ctx).Errorf("Error getting pullthroughBlobStore for image %q: %v", ref.Exact(), err)
+		}
+		return distribution.Descriptor{}, err
+	}
+
+	r.digestToStore[dgst.String()] = pullthroughBlobStore
+	return desc, nil
+}
+
+// ServeBlob attempts to serve the requested digest onto w, using a remote proxy store if necessary.
+func (r *pullthroughBlobStore) ServeBlob(ctx context.Context, w http.ResponseWriter, req *http.Request, dgst digest.Digest) error {
+	store, ok := r.digestToStore[dgst.String()]
+	if !ok {
+		return r.BlobStore.ServeBlob(ctx, w, req, dgst)
+	}
+
+	desc, err := store.Stat(ctx, dgst)
+	if err != nil {
+		context.GetLogger(r.repo.ctx).Errorf("Failed to stat digest %q: %v", dgst.String(), err)
+		return err
+	}
+
+	remoteReader, err := store.Open(ctx, dgst)
+	if err != nil {
+		context.GetLogger(r.repo.ctx).Errorf("Failure to open remote store %q: %v", dgst.String(), err)
+		return err
+	}
+
+	setResponseHeaders(w, desc.Size, desc.MediaType, dgst)
+
+	context.GetLogger(r.repo.ctx).Infof("Copying %d bytes of type %q for %q", desc.Size, desc.MediaType, dgst.String())
+	if _, err := io.CopyN(w, remoteReader, desc.Size); err != nil {
+		context.GetLogger(r.repo.ctx).Errorf("Failed copying content from remote store %q: %v", dgst.String(), err)
+		return err
+	}
+	return nil
+}
```

Comment:
We don't know layers are federated unless they have unambiguous metadata *inside* the layer.  We don't have enough of that info today to make a decision.  My concern is that what we're discussing there is not a solvable problem.